<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Biplabendu Das" />


<title>Locating harvester ant genes of interest in the daily gene expression network of carpenter ants</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-121215066-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-121215066-1');
</script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Biplabendu Das</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="research.html">
    <span class="fa fa-flask"></span>
     
    Research
  </a>
</li>
<li>
  <a href="camponotus_floridanus.html">
    <span class="fa fa-bug"></span>
     
    Applications
  </a>
</li>
<li>
  <a href="resume_10Mar23.pdf">
    <span class="fa fa-address-card"></span>
     
    Resume
  </a>
</li>
<li>
  <a href="ecology.html">
    <span class="fa fa-leaf"></span>
     
    Teaching
  </a>
</li>
<li>
  <a href="IUSSI2022.html">
    <span class="fa fa-pie-chart"></span>
     
    IUSSI
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://twitter.com/another_ant_guy">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://www.linkedin.com/in/biplabendu/">
    <span class="fa fa-linkedin"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/biplabendu">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Locating harvester ant genes of interest in the daily gene expression network of carpenter ants</h1>
<h4 class="author">Biplabendu Das</h4>
<h4 class="date">06 April, 2023</h4>

</div>


<div id="what-is-this-document" class="section level2">
<h2>What is this document?</h2>
<p>This document provides the method used in Das and Gordon (in prep) to answer the following question:</p>
<blockquote>
<p>Given genes of interest in <em>Pogonomyrmex barbatus</em>, for which temporal gene expression data is unavailable, can we infer if the genes of interest are involved in clock-controlled rhythmic processes?</p>
</blockquote>
<p>The methods to perform the following steps were adopted from de <a href="https://onlinelibrary.wiley.com/doi/full/10.1111/pim.12909">Bekker and Das (2022)</a> - build a circadian gene co-expression network (GCN), - how to annotate the network using published data, - infer functions of your gene-clusters-of-interest.</p>
</div>
<div id="step-1-build-circadian-gcn" class="section level2">
<h2>Step 1: Build circadian GCN</h2>
<div id="load-data" class="section level3">
<h3>1.1 Load data</h3>
<blockquote>
<p>For the carpenter ant, <em>Camponotus floridanus</em>, <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-021-08282-x">Das and de Bekker (2022)</a> have sampled forager and nurse brains every 2h, over a 24h period. We will utilize their published dataset to build a gene co-expression network (GCN) of ant brains.</p>
</blockquote>
<p>The raw data is deposited in NCBI under BioProject PRJNA704762.</p>
<p>Description of the dataset: Three forager and three nurse ant brains were sampled and pooled for RNA extraction and Illumina sequencing, every 2h over a 24h period. This resulted in 24 RNASeq datasets for ant brains (12 forager and 12 nurse datasets over the course of a 24h LD 12:12 day).</p>
<p>One would need to perform the usual steps – trimming the reads, mapping the reads to the genome, and quantifying normalized gene counts – to obtain normalized gene expression data from the raw reads. At the end, for each gene in the genome, we should have the normalized expression for each time point, throughout the 24h day.</p>
<p>For the purpose of this tutorial, we assume that you have organized the processed data into a (gene-expr X time-point) format, in a chronological order.</p>
<p>X2F = forager brain sampled at ZT2 (2h after lights were turned on), X4F = forager brain sampled at ZT4, and so on.</p>
<p>Now we read the data into R.</p>
<pre class="r"><code># loading database which contains data for Das and de Bekker 2021 (bioRxiv)
db &lt;- dbConnect(RSQLite::SQLite(), paste0(path_to_repo,&quot;/data/databases/TC5_data.db&quot;))

# extract the (gene-expr X time-point) data
dat &lt;-
  db %&gt;%
  tbl(., &quot;annot_fpkm&quot;) %&gt;%
  ## only forager expression data
  # select(gene_name, X2F:X24F) %&gt;%
  ## forager and nurse expression data
  select(gene_name, X2F:X24N) %&gt;%
  collect()

dim(dat)</code></pre>
<pre><code>## [1] 13813    25</code></pre>
</div>
<div id="clean-data" class="section level3">
<h3>1.2 Clean data</h3>
<p>The above dataset contains all genes (n=13,813) in the ant genome. However, not all of these genes are expressed in the ant brain, and some are expressed at very low levels that are not biologically meaningful.</p>
<p>Therefore, we will only keep the genes that are “expressed” (≥1 FPKM) in the ant brain, for at least half of all the sampled time points.</p>
<pre class="r"><code># Which genes are expressed throughout the day in both forager and nurses brains?
daily.exp.genes &lt;-
  tbl(db, &quot;expressed_genes&quot;) %&gt;% # note, the information is already available in the database
  filter(exp_half_for == &quot;yes&quot; &amp; exp_half_nur == &quot;yes&quot;) %&gt;%
  collect() %&gt;%
  pull(gene_name)

# Subset the gene-expr X time-point file
dat &lt;- dat %&gt;% filter(gene_name %in% daily.exp.genes)
dim(dat)</code></pre>
<pre><code>## [1] 9139   25</code></pre>
<p>This is our cleaned, input data file. The daily expression for these 9139 genes will be used to create the circadian GCN of Camponotus floridanus.</p>
</div>
<div id="format-data" class="section level3">
<h3>1.3 Format data</h3>
<p>To create the ant GCN, we will need to calculate the expression similarity (co-expression) of different gene pairs. Therefore, we would like to normalize the gene expression data by log2-transformation. Let’s do that and visualize the result.</p>
<pre class="r"><code>datExpr = as.data.frame(t(log2(dat[-c(1)]+1)))
names(datExpr) = dat$gene_name
rownames(datExpr) = names(dat)[-c(1)]

# ----------------------------------------------------------- #
# USE THE FOLLOWING CODE TO CHECK IF YOU HAVE ANY BAD SAMPLES #
# ----------------------------------------------------------- #
  # gsg = goodSamplesGenes(datExpr0, verbose = 3);
  # gsg$allOK
  #
  # sampleTree = hclust(dist(datExpr0), method = &quot;average&quot;);
  # # Plot the sample tree: Open a graphic output window of size 12 by 9 inches
  # # The user should change the dimensions if the window is too large or too small.
  # sizeGrWindow(12,9)
  # #pdf(file = &quot;Plots/sampleClustering.pdf&quot;, width = 12, height = 9);
  # par(cex = 1);
  # par(mar = c(0,4,2,0))
  # plot(sampleTree, main = &quot;Sample clustering to detect outliers&quot;, sub=&quot;&quot;, xlab=&quot;&quot;, cex.lab = 1.5,
  #      cex.axis = 1.5, cex.main = 2)
# ----------------------------------------------------------- #

# save the number of genes and samples
# that will be used to create the circadian GCN
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)

# visualize the log-transformed data
x = reshape2::melt(as.matrix(t(datExpr)))
colnames(x) = c(&#39;gene_id&#39;, &#39;sample&#39;, &#39;value&#39;)
ggplot(x, aes(x=value, color=sample)) + geom_density() + theme_Publication()</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/format_data-1.png" width="672" /></p>
<blockquote>
<p>Figure 1. Normalized gene expression: The density plot above shows the distribution of log2-transformed gene expression values for each sample.</p>
</blockquote>
</div>
<div id="calculate-gene-gene-similarity" class="section level3">
<h3>1.4 Calculate gene-gene similarity</h3>
<p>Now, we can calculate the pairwise gene expression similarity for each of the 9139 genes and save it to a matrix.</p>
<p>I calculated expression similarity for all gene pairs in a dataset using Kendall’s tau, which measures the ordinal relationship between two variables and is used in rhythmicity detection algorithms [1].</p>
<pre class="r"><code># Calculate Kendall&#39;s tau-b correlation for each gene-gene pair

# sim_matrix &lt;- cor((datExpr), method = &quot;kendall&quot;) # this step takes time
# save(sim_matrix, file = paste0(path_to_repo, &quot;/results/temp_files/sim_matrix_for_nur_TC5.RData&quot;)) # might be useful to save the sim_matrix and
load(paste0(path_to_repo, &quot;/results/temp_files/sim_matrix_for_nur_TC5.RData&quot;)) # load it up
# save(sim_matrix, file = paste0(path_to_repo, &quot;/results/temp_files/sim_matrix_for_TC5.RData&quot;)) # might be useful to save the sim_matrix and
# load(paste0(path_to_repo, &quot;/results/temp_files/sim_matrix_for_TC5.RData&quot;)) # load it up

## Let&#39;s display a chunk of the matrix (code from Hughitt 2016; github)
heatmap_indices &lt;- sample(nrow(sim_matrix), 500)
gplots::heatmap.2(t(sim_matrix[heatmap_indices, heatmap_indices]),
          col=inferno(100),
          labRow=NA, labCol=NA,
          trace=&#39;none&#39;, dendrogram=&#39;row&#39;,
          xlab=&#39;Gene&#39;, ylab=&#39;Gene&#39;,
          main=&#39;Similarity matrix \n correlation method = &quot;kendall&quot; \n (500 random genes)&#39;,
          density.info=&#39;none&#39;, revC=TRUE)</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/gene_sim_matrix-1.png" width="672" /></p>
<blockquote>
<p>Figure 2. Similarity matrix: The heatmap shows the pairwise Kendall’s tau correlation for a set of 500 genes randomly pulled from the 9139 genes expressed in the ant brain.</p>
</blockquote>
</div>
<div id="create-adjacency-matrix" class="section level3">
<h3>1.5 Create adjacency matrix</h3>
<p>From the above similarity matrix, we then need to create the adjacency matrix needed for constructing a gene co-expression network.</p>
<p>To create the adjacency matrix, we need to first identify the soft-thresholding power by calling the network topology analysis function from the WGCNA package [2].</p>
<pre class="r"><code># Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to=20, by=2))
# # Call the network topology analysis function
sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)</code></pre>
<pre><code>## pickSoftThreshold: will use block size 4895.
##  pickSoftThreshold: calculating connectivity for given powers...
##    ..working on genes 1 through 4895 of 9139
##    ..working on genes 4896 through 9139 of 9139
##    Power SFT.R.sq  slope truncated.R.sq mean.k. median.k. max.k.
## 1      1    0.845  1.900          0.995  3310.0   3390.00   4730
## 2      2    0.248  0.276          0.930  1720.0   1710.00   3200
## 3      3    0.343 -0.284          0.907  1050.0    988.00   2410
## 4      4    0.696 -0.580          0.922   701.0    616.00   1930
## 5      5    0.818 -0.762          0.951   499.0    402.00   1600
## 6      6    0.847 -0.896          0.942   371.0    272.00   1360
## 7      7    0.854 -0.992          0.933   285.0    190.00   1180
## 8      8    0.868 -1.060          0.935   225.0    136.00   1030
## 9      9    0.879 -1.110          0.940   181.0     99.10    919
## 10    10    0.874 -1.160          0.928   148.0     73.40    824
## 11    12    0.879 -1.220          0.928   103.0     42.20    676
## 12    14    0.879 -1.280          0.921    74.8     25.40    568
## 13    16    0.874 -1.310          0.916    56.1     15.80    485
## 14    18    0.842 -1.360          0.884    43.2     10.10    420
## 15    20    0.827 -1.390          0.874    34.0      6.64    367</code></pre>
<pre class="r"><code># Plot the results:
# sizeGrWindow(9, 5)
par(mfrow = c(1,2));
cex1 = 0.9;
# Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab=&quot;Soft Threshold (power)&quot;,ylab=&quot;Scale Free Topology Model Fit,signed R^2&quot;,type=&quot;n&quot;,
     main = paste(&quot;Scale independence&quot;));
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col=&quot;red&quot;);
# this line corresponds to using an R^2 cut-off of h
abline(h=0.90,col=&quot;red&quot;)
# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab=&quot;Soft Threshold (power)&quot;,ylab=&quot;Mean Connectivity&quot;, type=&quot;n&quot;,
     main = paste(&quot;Mean connectivity&quot;))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col=&quot;red&quot;)</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/soft_thresholding_power-1.png" width="672" /></p>
<blockquote>
<p>Figure 3. Soft-thresholding power: The above plots show the effect of soft-thresholding power on the topology and the mean connectivity of the transformed similarity matrix (network).</p>
</blockquote>
<p>NOTE: The scale-free topology fit index reaches ~0.9 at a soft-thresholding power of 9 and it does not improve drastically beyond that.</p>
<p>So, we will set our soft thresholding power to 9 for creating the adjacency matrix.</p>
<pre class="r"><code>## Specify the soft-thresholding-power
soft.power = 9

# Construct adjacency matrix
adj_matrix &lt;- adjacency.fromSimilarity(sim_matrix,
                                       power=soft.power,
                                       type=&#39;signed&#39;
                                        )
save(adj_matrix, file = paste0(path_to_repo, &quot;/results/temp_files/adj_matrix_for_nur_TC5.RData&quot;)) # might be useful to save the sim_matrix and
# save(adj_matrix, file = paste0(path_to_repo, &quot;/results/temp_files/adj_matrix_for_TC5.RData&quot;)) # might be useful to save the sim_matrix and
# load(paste0(path_to_repo, &quot;/results/temp_files/adj_matrix_for_TC5.RData&quot;)) # load it up


# Convert adj_matrix to matrix
gene_ids &lt;- rownames(adj_matrix)

adj_matrix &lt;- matrix(adj_matrix, nrow=nrow(adj_matrix))
rownames(adj_matrix) &lt;- gene_ids
colnames(adj_matrix) &lt;- gene_ids

## Same heatmap as before, but now with the power-transformed adjacency matrix
gplots::heatmap.2(t(adj_matrix[heatmap_indices, heatmap_indices]),
                  col=inferno(100),
                  labRow=NA, labCol=NA,
                  trace=&#39;none&#39;, dendrogram=&#39;row&#39;,
                  xlab=&#39;Gene&#39;, ylab=&#39;Gene&#39;,
                  main=&#39;Adjacency matrix&#39;,
                  density.info=&#39;none&#39;, revC=TRUE)</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/adjacency_matrix-1.png" width="672" /></p>
<p>Figure 4. Adjacency matrix: The heatmap shows the result of the power-transformation on the similarity of the 500 random genes shown previously in Figure 2. As you can see, only the highest pair-wise correlations are retained whereas the weak correlations tend to zero.</p>
<hr />
</div>
</div>
<div id="step-2-identify-gene-clusters" class="section level2">
<h2>Step 2: Identify gene clusters</h2>
<div id="create-topological-overalp-matrix" class="section level3">
<h3>2.1 Create topological overalp matrix</h3>
<pre class="r"><code># Turn adjacency into topological overlap
# TOM = TOMsimilarity(adj_matrix);
# dissTOM = 1-TOM
# save(dissTOM, file = paste0(path_to_repo, &quot;/results/temp_files/dissTOM_for_nur_TC5.RData&quot;)) # might be useful to save the sim_matrix and
# save(dissTOM, file = paste0(path_to_repo, &quot;/results/temp_files/dissTOM_for_TC5.RData&quot;)) # might be useful to save the sim_matrix and
load(paste0(path_to_repo, &quot;/results/temp_files/dissTOM_for_nur_TC5.RData&quot;)) # load it up
# load(paste0(path_to_repo, &quot;/results/temp_files/dissTOM_for_TC5.RData&quot;)) # load it up

# Call the hierarchical clustering function
geneTree = hclust(as.dist(dissTOM), method = &quot;average&quot;)

# Plot the resulting clustering tree (dendrogram)
# sizeGrWindow(12,9)
plot(geneTree, xlab=&quot;&quot;, sub=&quot;&quot;, main = &quot;Gene clustering on TOM-based dissimilarity&quot;,
     labels = FALSE, hang = 0.04)</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/adj_to_TOM-1.png" width="672" /></p>
<blockquote>
<p>Figure 5. Clustering tree: The figure shows the clustering tree (dendrogram) that results from hierarchical clustering of the TOM-based dissimilarity matrix and will be used for identifying modules of highly similar genes in the co-expression network.</p>
</blockquote>
</div>
<div id="identify-clusters" class="section level3">
<h3>2.2 Identify clusters</h3>
<p>To cluster genes with similar daily expression pattern, we use the cutreeDynamic() function from the WGCNA package.</p>
<p>We need to provide a minimum size for the identified clusters or modules. This can set depending on the user’s question. In our case, we want to identify fairly large modules that are biologically meaningful (i.e., enriched in different GO/PFAM terms). As such, we set the minimum module size to 30. However, as you will see later, we will refine our cluster identification by merging very similar modules. As such, the choice of minimum module size should not affect cluster identification drastically.</p>
<pre class="r"><code># We like large modules, so we set the minimum module size relatively high:
minModuleSize = 30;

# Module identification using dynamic tree cut:
dynamicMods= cutreeDynamic(dendro = geneTree,
                           distM = dissTOM,
                           method = &quot;hybrid&quot;,
                           verbose = 4,
                           deepSplit = 3, # see WGCNA for more info on tuning parameters
                           pamRespectsDendro = FALSE,
                           minClusterSize = minModuleSize);</code></pre>
<pre><code>##  ..cutHeight not given, setting it to 0.99  ===&gt;  99% of the (truncated) height range in dendro.
##  ..Going through the merge tree
##  
##  ..Going through detected branches and marking clusters..
##  ..Assigning Tree Cut stage labels..
##  ..Assigning PAM stage labels..
##  ....assigned 5531 objects to existing clusters.
##  ..done.</code></pre>
<pre class="r"><code># view number of genes in each module
table(dynamicMods)</code></pre>
<pre><code>## dynamicMods
##    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16 
## 1337 1172  742  723  720  579  515  506  369  299  269  265  199  149  134  134 
##   17   18   19   20   21   22   23   24   25   26   27   28   29   30 
##  112  107   99   91   88   79   75   72   66   56   55   50   45   32</code></pre>
<pre class="r"><code># Convert numeric lables into colors
dynamicColors = labels2colors(dynamicMods)
table(dynamicColors)</code></pre>
<pre><code>## dynamicColors
##         black          blue         brown          cyan     darkgreen 
##           515          1172           742           149            79 
##      darkgrey    darkorange       darkred darkturquoise         green 
##            72            56            88            75           720 
##   greenyellow        grey60     lightcyan    lightgreen   lightyellow 
##           269           112           134           107            99 
##       magenta  midnightblue        orange          pink        purple 
##           369           134            66           506           299 
##           red     royalblue   saddlebrown        salmon       skyblue 
##           579            91            45           199            50 
##     steelblue           tan     turquoise         white        yellow 
##            32           265          1337            55           723</code></pre>
<p><strong>In the initial cluster (module) identification step, WGCNA finds 30 modules. However, some of the identified modules might have very similar expression pattern and we would rather merge this closely related modules into one.</strong></p>
<p>We do that in the next step.</p>
</div>
<div id="merge-similar-modules" class="section level3">
<h3>2.3 Merge similar modules</h3>
<pre class="r"><code># Calculate eigengenes
MEList = moduleEigengenes(datExpr, colors = dynamicColors)
MEs = MEList$eigengenes

# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs, method = &quot;kendall&quot;);

# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = &quot;average&quot;);
# Plot the result
# sizeGrWindow(7, 8)
plot(METree, main = &quot;Clustering of module eigengenes&quot;,
     xlab = &quot;&quot;, sub = &quot;MEDiss = 1-cor(MEs, method = &#39;kendall&#39;)&quot;)

# We choose a height cut of 0.2, corresponding to correlation of 0.8, to merge
MEDissThres = 0.2 # user-specified parameter value; see WGCNA manual for more info

# Plot the cut line into the dendrogram
abline(h=MEDissThres, col = &quot;red&quot;)</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/refine_cluster_part1-1.png" width="672" /></p>
<blockquote>
<p>Figure 6. Merging similar modules: The above figure shows the similarity of the different gene modules using hierarchical clustering of the module’s eigenvalue (eigengene expression). The horizontal red line shows the cutoff used to merge similar modules.</p>
</blockquote>
<p>We choose a cut height of 0.2, corresponding to correlation of 0.8, to merge similar modules. Although arbitrary, the cutoff was motivated by the number of modules we would like to retain in the GCN; in our case, a 0.2 threshold resulted in a total of 12 modules in the GCN (see below).</p>
<p>In the following code, we merge the similar modules and visualize the module assignments before and after merging.</p>
<pre class="r"><code># Call an automatic merging function
merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = 3)</code></pre>
<pre><code>##  mergeCloseModules: Merging modules whose distance is less than 0.2
##    multiSetMEs: Calculating module MEs.
##      Working on set 1 ...
##      moduleEigengenes: Calculating 30 module eigengenes in given set.
##    multiSetMEs: Calculating module MEs.
##      Working on set 1 ...
##      moduleEigengenes: Calculating 13 module eigengenes in given set.
##    multiSetMEs: Calculating module MEs.
##      Working on set 1 ...
##      moduleEigengenes: Calculating 12 module eigengenes in given set.
##    Calculating new MEs...
##    multiSetMEs: Calculating module MEs.
##      Working on set 1 ...
##      moduleEigengenes: Calculating 12 module eigengenes in given set.</code></pre>
<pre class="r"><code># The merged module colors
mergedColors = merge$colors;
# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs;

# sizeGrWindow(12, 9)
plotDendroAndColors(geneTree,
                    cbind(dynamicColors, mergedColors),
                    c(&quot;Dynamic Tree Cut&quot;, &quot;Merged dynamic&quot;),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/refine_cluster_part2-1.png" width="672" /></p>
<pre class="r"><code># Rename to moduleColors
moduleColors = mergedColors

# Construct numerical labels corresponding to the colors
colorOrder = c(&quot;grey&quot;, standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1</code></pre>
<blockquote>
<p>Figure 7. Modules of highly co-expressed genes: The above plot shows the results of module identification before (Dyanamic Tree Cut) and after (Merged dyanamic) similar modules were merged.</p>
</blockquote>
<p><strong>We identified 12 modules in the ant GCN, the size of each of these modules are shown below.</strong></p>
<p>Note, WGCNA names the different modules as colors (see above), and the colors have no meaning. Therefore, it might be useful to rename the modules. In the next step we rename all the modules according to the following convention:</p>
</div>
<div id="calculate-module-module-similarity" class="section level3">
<h3>2.4 Calculate module-module similarity</h3>
<p>Thus far, we have created the ant GCN (adjacency matrix) and identified 12 modules of highly co-expressed genes in the network.</p>
<p>Next, we investigate how the different modules are connected to each other in the GCN. To do so, we calculate the module-module similarity (Kendall’s tau-b correlation for pairwise module-eigengene expression) and then use the similarity matrix to create the module adjacency matrix.</p>
<p>The following code calculates the module adjacency matrix and visualizes it as a heatmap.</p>
<pre class="r"><code># Calculate similarity of the eigen-genes
sim_matrix_ME &lt;- cor(mergedMEs, method = &quot;kendall&quot;)

# calculate adj_matrix
adj_matrix_ME &lt;- adjacency.fromSimilarity(sim_matrix_ME,
                                          power=1, # DO NOT power transform
                                          type=&#39;signed&#39;
)

# coerce into a matrix

## GET THE NAMES OF THE MODULES
# module_ids &lt;- rownames(adj_matrix_ME)
## CHANGE THE NAMES OF THE MODULES
module_ids &lt;- data.frame(old_labels = rownames(adj_matrix_ME),
                       new_labels = paste0(&quot;module-&quot;, 1:nrow(adj_matrix_ME)))

adj_matrix_ME &lt;- matrix(adj_matrix_ME, nrow=nrow(adj_matrix_ME))
rownames(adj_matrix_ME) &lt;- module_ids$new_labels
colnames(adj_matrix_ME) &lt;- module_ids$new_labels

# png(paste0(path_to_repo, &quot;/results/figures/ME_adjacency.png&quot;), 
#     width = 18, height = 18, units = &quot;cm&quot;, res = 400)
gplots::heatmap.2(t(adj_matrix_ME),
                  col=inferno(100),
                  # labRow=NA, labCol=NA,
                  trace=&#39;none&#39;, dendrogram=&#39;row&#39;,
                  xlab=&#39;&#39;, ylab=&#39;&#39;,
                  # main=&#39;Similarity matrix - MEs \n correlation method = &quot;kendall&quot;)&#39;,
                  main=&#39;Adjacency matrix - MEs&#39;,
                  density.info=&#39;none&#39;, revC=TRUE)</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/module_sim_matrix-1.png" width="672" /></p>
<pre class="r"><code># trash &lt;- dev.off()</code></pre>
<blockquote>
<p>Figure 8. Module-module relationships: The heatmap above shows the pairwise Kendall’s tau correlation (similarity) of the twelve modules identified in the ant GCN. Darker shades indicate low correlations and brighter shades indicate high correlations, as shown in the Color Key.</p>
</blockquote>
</div>
<div id="visualize-the-network" class="section level3">
<h3>2.5 Visualize the network</h3>
<p>To better visualize the global network – how the modules are connected to each other – we can simplify the network. That is, we remove most of the weak edges of the network and retain only the strong module-module correlations.</p>
<p>For example, to remove weak edges, we can set all correlations less than 0.6 to be zero. This will help us obtain a fairly clean network for visualization. To simplify further, we can assign the same edge weight for all correlations between 0.6 and 0.8 (e.g., 0.5), and a different edge weight for correlations ≥ 0.8 (e.g., 1).</p>
<p>The following code uses the igraph package in R to simplify and visualize the module-module relationships in the network.</p>
<pre class="r"><code>pacman::p_load(igraph)

# get rid of low correlations (0.6 &amp; 0.8 are arbitrary)
adj_matrix_ME[adj_matrix_ME &lt; 0.6] &lt;- 0
adj_matrix_ME[adj_matrix_ME &lt; 0.8 &amp; adj_matrix_ME&gt;0] &lt;- 0.5
adj_matrix_ME[adj_matrix_ME &gt;= 0.8] &lt;- 1

# build_network
network &lt;- graph.adjacency(adj_matrix_ME,
                           mode = &quot;upper&quot;,
                           weighted = T,
                           diag = F)

# simplify network
network &lt;- igraph::simplify(network)  # removes self-loops

# E(network)$width &lt;- E(network)$weight + min(E(network)$weight) + 1 # offset=1

colors &lt;- mergedMEs %&gt;% names() %&gt;% str_split(&quot;ME&quot;, 2) %&gt;% sapply(&quot;[&quot;, 2)
V(network)$color &lt;- colors

genes_ME &lt;- factor(moduleColors, levels=colors) %&gt;% summary()
V(network)$size &lt;- log2(genes_ME)*2

V(network)$label.color &lt;- &quot;black&quot;
V(network)$frame.color &lt;- &quot;white&quot;

E(network)$width &lt;- E(network)$weight^2*4
E(network)$edge.color &lt;- &quot;gray80&quot;

# par(mar=c(0,0,0,0))
# remove unconnected nodes
# network &lt;- delete.vertices(network, degree(network)==0)
# plot(network,
#      layout=layout.fruchterman.reingold
#      # layout = layout.kamada.kawai
#      # layout = layout.kamada.kawai
#      )

# trash &lt;- dev.off()
## Circular layout
plot(network,
     layout=layout.kamada.kawai,
       # layout=layout.fruchterman.reingold
       # layout=layout.graphopt
       # layout=layout_in_circle,
     # vertex.label=NA
     # vertex.size=hub.score(network)$vector*30
     vertex.shape=&quot;none&quot;
)</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/visualize_network-1.png" width="576" /></p>
<blockquote>
<p>Figure 9. Visualizing the ant GCN: A simplified view of the connectivity patterns between the different gene modules of the ant GCN are shown. In our case, thick edges between two modules indicate correlations ≥ 0.8, thinner edges indicate correlations between (0.6, 0.8), and no edges indicate correlations &lt; 0.6.</p>
</blockquote>
</div>
</div>
<div id="step-3-define-genes-of-interest" class="section level2">
<h2>Step 3: Define genes of interest</h2>
<div id="module-gene-identity" class="section level3">
<h3>3.1 Module-gene identity</h3>
<p>Split all the genes into different sets, one for each module</p>
<pre class="r"><code># DEFINE GENES OF INTEREST (PART 3)

# Make a list that contains all gene names for a given cluster
module_color = colors
module = names(mergedMEs)
module_colors &lt;-
  data.frame(module_label=module) %&gt;%
  mutate(module_color = str_replace(module_label, &quot;ME&quot;, &quot;&quot;))

module_genes &lt;- list()
module_color &lt;- module_colors$module_color
# Get the genes from each of the modules
for (i in 1:length(module_color)) {

  module_genes[[i]] &lt;- names(datExpr)[which(moduleColors==module_color[[i]])]
  names(module_genes)[[i]] &lt;- module_color[[i]]
}

names(module_genes) &lt;- module_ids$new_labels

# LIST ONE - WGCNA modules
list1 &lt;- module_genes
sapply(list1, length)</code></pre>
<pre><code>##  module-1  module-2  module-3  module-4  module-5  module-6  module-7  module-8 
##       403        56       922      2269       127       179       664      2616 
##  module-9 module-10 module-11 module-12 
##       209        66      1533        95</code></pre>
</div>
<div id="camponotus-floridanus" class="section level3">
<h3>3.2 Camponotus floridanus</h3>
<p>We want to identify the GCN modules that contain</p>
<ul>
<li>24h oscillating genes in Cflo (for.rhy = 24h-rhythmic genes in forager brains, nur.rhy = 24h-rhythmic genes in nurses).</li>
</ul>
<pre class="r"><code># DEFINE GENES OF INTEREST (PART 1)

# load the data table that contains the results from the rhythmicity analysis (eJTK-Cycle output)
rhy.trait.24 &lt;- tbl(db, &quot;ejtk_all&quot;) %&gt;% select(gene_name:rhy) %&gt;% collect()

# pull the gene names that are rhythmic in forager brains
for.rhy &lt;- rhy.trait.24 %&gt;% filter(caste==&quot;for&quot; &amp; rhy==&quot;yes&quot;) %&gt;% pull(gene_name)
# pull the gene names that are rhythmic in forager brains
nur.rhy &lt;- rhy.trait.24 %&gt;% filter(caste==&quot;nur&quot; &amp; rhy==&quot;yes&quot;) %&gt;% pull(gene_name)
# genes that show 24h rhythms in forager brains and 8h rhythms in nurses
for24.nur8 &lt;- 
  tbl(db, &quot;ejtk_8h_all&quot;) %&gt;% 
  filter(caste==&quot;nur&quot; &amp; rhy==&quot;yes&quot;) %&gt;% collect() %&gt;% pull(gene_name) %&gt;% 
  intersect(., for.rhy) %&gt;% unique()

# DEFINE GENES OF INTEREST (PART 2)

## Genes underlying behavioral plasticity, i.e., DEGS (foragers v. nurses)
  # genes higher expressed in forager brains (v. nurse brains)
  for.up &lt;- tbl(db, &quot;TC5_DEGs_all&quot;) %&gt;% filter(upregulation==&quot;for&quot;) %&gt;% collect() %&gt;% pull(gene_name)
  # genes lower expressed in for. brains (v. nurse brains)
  for.down &lt;- tbl(db, &quot;TC5_DEGs_all&quot;) %&gt;% filter(upregulation==&quot;nur&quot;) %&gt;% collect() %&gt;% pull(gene_name)</code></pre>
</div>
</div>
<div id="pogonomyrmex-barbatus" class="section level2">
<h2>3.2 Pogonomyrmex barbatus</h2>
<div id="pogo-cflo-orthologs" class="section level3">
<h3>3.2.1 Pogo-Cflo orthologs</h3>
<p>Read the one-to-one orthology data for P. barbatus and C. floridanus. The data was obtained by using the program proteinortho.</p>
<pre class="r"><code>pogo.cflo &lt;- read.csv(paste0(path_to_pogodata, &quot;pogo_cflo_orthologs.csv&quot;),
                      header = T, stringsAsFactors = F, na.strings = c(&quot;&quot;,&quot; &quot;,&quot;.&quot;, &quot;NA&quot;)) %&gt;% as_tibble()</code></pre>
</div>
<div id="response-to-humidity" class="section level3">
<h3>3.2.2 Response to humidity</h3>
<ul>
<li>Read the data</li>
<li>Set correlation threshold | cor(collective behav, geneexpression) = [-0.6,0.5]</li>
<li>we selected the top 5% as the positively correlated genes</li>
<li>similarly, bottom 5% as the negetively</li>
<li>Note, we tried using higher thresholds, but the results do not change vastly.</li>
</ul>
<pre class="r"><code># Read 2020_Friedman_et_al data -------------------------------------------

df.2020 &lt;- read.csv(paste0(path_to_pogodata, &quot;2020_friedman_Pbar_TPM_TraitCorr_dNdS.csv&quot;), 
                    header = T, stringsAsFactors = F, na.strings = c(&quot;&quot;,&quot; &quot;,&quot;.&quot;, &quot;NA&quot;)) %&gt;% as_tibble()

# note, it is unclear what the columns actually mean, 
# but, below find an example of what we could do.
# Assumption: &quot;humidity_cor&quot; represents the correlation of gene expression to humidity changes

## Keep relevant columns
df.2020 &lt;- 
  df.2020 %&gt;% 
  select(pogo_gene = LOC,
         humidity_cor:da_5ht_cor_colony_mean)

## Deciding on the threshold
## the first and the third quartiles

# Set humidity-expression correlation threshold
h.threshold &lt;- quantile(df.2020$humidity_cor, prob=c(.05,.95), type=1)

# obtain pogo genes that satisfied the threshold
h.genes.pos &lt;- 
  df.2020 %&gt;% 
  as_tibble() %&gt;% 
  select(pogo_gene,
         humidity_cor) %&gt;% 
  na.omit() %&gt;% 
  filter(humidity_cor &gt; h.threshold[2]) %&gt;% 
  pull(pogo_gene)
h.genes.neg &lt;- 
  df.2020 %&gt;% 
  as_tibble() %&gt;% 
  select(pogo_gene,
         humidity_cor) %&gt;% 
  na.omit() %&gt;% 
  filter(humidity_cor &lt; h.threshold[1]) %&gt;% 
  pull(pogo_gene)
# get cflo orthologs for these genes
h.genes.cflo.pos &lt;-
  pogo.cflo %&gt;% 
  select(pogo_gene, cflo_gene) %&gt;% 
  distinct() %&gt;% 
  filter(pogo_gene %in% h.genes.pos) %&gt;% 
  pull(cflo_gene) %&gt;% unique()
h.genes.cflo.neg &lt;-
  pogo.cflo %&gt;% 
  select(pogo_gene, cflo_gene) %&gt;% 
  distinct() %&gt;% 
  filter(pogo_gene %in% h.genes.neg) %&gt;% 
  pull(cflo_gene) %&gt;% unique()</code></pre>
<blockquote>
<p><code>471 of the 516</code> negatively correlated pogo genes had an ortholog in Cflo, whereas <code>441 of the 516</code> positivey correlated pogo genes had an ortholog in Cflo.</p>
</blockquote>
</div>
<div id="annotate-the-network" class="section level3">
<h3>4.4 Annotate the network</h3>
<blockquote>
<p>Now, let’s see which modules, if any, in the cflo brain GCN do these genes reside?</p>
</blockquote>
<p>The names have the following meanings:</p>
<ul>
<li>for-UP = genes upregulated in forager brains as compared to nurses</li>
<li>for-DOWN = genes downregulated in forager brains as compared to nurses</li>
<li>for-rhy24 = genes that show 24h-rhythmic expression in forager brains</li>
<li>for24-nur8 = genes that show 24h-rhythmic expression in forager brains, but 8h rhythms in nurses</li>
<li>pogo-rH-pos = Cflo-orthologs of pogonomyrmex genes whose expression show a positive correlation with collective response to abiotic changes</li>
<li>pogo-rH-neg = Cflo-orthologs of pogonomyrmex genes whose expression show a negative correlation with collective response to abiotic changes</li>
</ul>
<pre class="r"><code>## LIST THREE - genes underlying behavioral plasticity and parasitic behavioral manipulation
list4 &lt;- list(for.up, for.down, # same as list three
              for.rhy, for24.nur8,
              h.genes.cflo.pos, h.genes.cflo.neg)
names(list4) &lt;- c(&quot;for-UP&quot;, &quot;for-DOWN&quot;,
                  &quot;for-rhy24&quot;, &quot;for24-nur8&quot;,
                  &quot;pogo-rH-pos&quot;, &quot;pogo-rH-neg&quot;)

## CHECK FOR OVERLAP

## what is my background space? In other words, how many pogo genes have a cflo ortholog?
nGenes.pogo &lt;- pogo.cflo %&gt;% pull(cflo_gene) %&gt;% unique() %&gt;% length()

## make a GOM object
gom.1v4 &lt;- newGOM(list1, list4,
       genome.size = nGenes.pogo)

drawHeatmap(gom.1v4,
              adj.p=T,
              cutoff=0.01,
              what=&quot;odds.ratio&quot;,
              # what=&quot;Jaccard&quot;,
              log.scale = T,
              note.col = &quot;grey60&quot;)</code></pre>
<p><img src="comparing_cflo_pogo_gordonlab_files/figure-html/pogo_module-1.png" width="864" /></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
